client/block.c

```diff
@@ -921,74 +923,97 @@ int xdag_create_block(struct xdag_field *fields, int inputsCount, int outputsCou
-		pthread_mutex_unlock((pthread_mutex_t*)g_ptr_share_mutex);
+	block->field[0].transport_header = 1;
+	int res = xdag_add_block(block);
+	if(res > 0) {
+		xdag_send_new_block(block);
+		res = 1;
+	}
+	else {
+		res = 0;
 	}
+	free(block);
 
-	xdag_hash(block, sizeof(struct xdag_block), newBlockHash);
-	block[0].field[0].transport_header = 1;
+	return res;
+}
 
-	log_block("Create", newBlockHash, block[0].field[0].time, 1);
+int do_mining(struct xdag_block *block, struct block_internal *pretop, xdag_time_t send_time)
+{
+	uint64_t taskIndex = g_xdag_pool_task_index + 1;
+	struct xdag_pool_task *task = &g_xdag_pool_task[taskIndex & 1];
 
-	res = xdag_add_block(block);
-	if (res > 0) {
-		if (mining) {
-			memcpy(g_xdag_mined_hashes[MAIN_TIME(send_time) & (CONFIRMATIONS_COUNT - 1)],
-				newBlockHash, sizeof(xdag_hash_t));
-			memcpy(g_xdag_mined_nonce[MAIN_TIME(send_time) & (CONFIRMATIONS_COUNT - 1)],
-				block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(xdag_hash_t));
-		}
+	xdag_generate_random_array(block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(xdag_hash_t));
 
-		if(g_xdag_pool) { /* append pool created block to list */
-			struct xdag_block * new_block = (struct xdag_block *)malloc(sizeof(struct xdag_block));
-			memcpy(new_block, block, sizeof(struct xdag_block));
-			new_block->field[0].transport_header = 0;
-			xdag_append_new_block(new_block);
-		} else { /* send miner created block directly */
-			xdag_send_new_block(block);
-		}
+	task->task_time = MAIN_TIME(send_time);
 
-		if(newBlockHashResult != NULL) {
-			memcpy(newBlockHashResult, newBlockHash, sizeof(xdag_hash_t));
+	xdag_hash_init(task->ctx0);
+	xdag_hash_update(task->ctx0, block, sizeof(struct xdag_block) - 2 * sizeof(struct xdag_field));
+	xdag_hash_get_state(task->ctx0, task->task[0].data);
+	xdag_hash_update(task->ctx0, block[0].field[XDAG_BLOCK_FIELDS - 2].data, sizeof(struct xdag_field));
+	memcpy(task->ctx, task->ctx0, xdag_hash_ctx_size());
+
+	xdag_hash_update(task->ctx, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field) - sizeof(uint64_t));
+	memcpy(task->task[1].data, block[0].field[XDAG_BLOCK_FIELDS - 2].data, sizeof(struct xdag_field));
+	memcpy(task->nonce.data, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field));
+	memcpy(task->lastfield.data, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field));
+
+	xdag_hash_final(task->ctx, &task->nonce.amount, sizeof(uint64_t), task->minhash.data);
+	g_xdag_pool_task_index = taskIndex;
+
+	while(get_timestamp() <= send_time) {
+		sleep(1);
+		struct block_internal *pretop_new = pretop_block();
+		if(pretop != pretop_new && get_timestamp() < send_time) {
+			pretop = pretop_new;
+			xdag_info("Mining: start from beginning because of pre-top block changed");
+			return 0;
 		}
-		res = 0;
 	}
 
-	return res;
-}
+	pthread_mutex_lock((pthread_mutex_t*)g_ptr_share_mutex);
```

Comment:
where this lock is also used? i can't find it

Reply:
It's a mutex for both miner and pool as far as I remember.

Reply:
This lock is not new, code section was just moved. It is necessary to protect the best found share between two threads.

Reply:
I know that it is not new, but i don't understand the logic of use a mutex to read task->lastfield.data while you write task->lastfield.data without this lock so i'm just asking what's the meaning of it
maybe i don't see where else this lock is used, i only see it here to read task->lastfield.data